### PR TITLE
Connection handling fixes

### LINKED ISSUE #7242: Improved efficiency of ConnectionHandlingMode: deeper integration between Hibernate ORM and Agroal
After having switched the Hibernate ORM's `PhysicalConnectionHandlingMode` to `DELAYED_ACQUISITION_AND_RELEASE_AFTER_TRANSACTION` as I believed it would have been a more efficient choice, I had to revert it as short term fix for #6770 . This is to track the propert fix.

**The problem**
Hibernate ORM's is assuming it deals with the "physical" connection, as the name of the property suggests, but when there is a running transaction Agroal doesn't return the raw connection but a wrapper.
At transaction commit, Agroal verifies that all wrappers which depend on the currenty physical connection have been closed first - otherwise it's assuming one has "escaped" the scope of its indended lifespan and it logs a warning about a possible leak.

At high level, assuming a JTA environment (like Quarkus _normally_ is):

1. a JTA transaction is started
2. a connection is acquired from the pool, so it needs to be enlisted in the existing TX
3. connection is being used by the application (Hibernate ORM in this case)
4. TX is committed

In such a scenario, if Hibernate's EntityManager [or Session] isn't closed before the TX is committed, the connection wrapper will still be open - hence the warning log.
A simple way forward is to expect the EntityManager to have been closed before commit, but this is counter-intuitive if we allow code looking like:

        EntityManager em = emf.createEntityManager();
        EntityTransaction transaction = em.getTransaction();
        transaction.begin();
        em.createNativeQuery("Delete from Person").executeUpdate();
        transaction.commit();
        em.close();

One problem of such code is that we seem to allow `em.getTransaction()`, which is normally not allowed in JTA (it's supposed to throw an `IllegalStateException`) as one should either let the container manage transactions, or explicitly request a reference to a UserTransaction. Clearly a benefit of letting the container handle such details is that it can control the lifecycle of the various parties and cleanup stuff in some controlled order.

## Question

What should `EntityManager.getTransaction()` do ? Is the current semantics reasonable?

N.B. if I understood the Agroal experts correctly, a single Agroal connection pool is either configured to integrate with JTA or not: if JTA integration is enabled, it will require a running transaction for any new connection request.

Also Jesper reminds: "we can't support deferred enlistement, like JCA can"

## Some Ideas

Some notes from previous brainstorming:

#### Implement a new DELAYED_ACQUISITION_AND_RELEASE_BEFORE_TRANSACTION

We could have Hibernate ORM release the connection just before the commit. 

@barreiro already created a branch: https://github.com/hibernate/hibernate-orm/compare/master...barreiro:before-completion

This seems simple and would be rather effective, but @jesperpedersen is warnings us against the use of interposed Synchronization(s); in WildFly for example they had to guarantee ordering to avoid some more complex problems with them.

[Q: is the use of Synchronization really a problem? ORM is already registering one, it seems Luis's code is merely using the existing synchronization]

#### Allow Hibernate ORM to bypass Agroal's safety net.

I'm not suggesting we should, but it would be interesting to consider this: both Agroal and Hibernate ORM seem to implement strict measures against leaking of resources. For sake of performance we should allow the two components to trust each other, and avoid duplicating checks.

It's tempting to have ORM access the real "physical" connection and:

- bypass the wrapper creation
- bypass the resource tracking of statements provided by Agroal
- as a consequence, no warnings about the leaked wrappers would be logged.

[I'm aware that resource tracking of statements can be disabled - but if we disable it on the pool, then it's also disabled for user code using the pool directly, which is likely not a good idea]

I suspect the strongest argument against this would be that Agroal itself would need to expose an unsafe API. Maybe not my best idea? But it would be nice to allow a "friend framework" some deeper integration for sake of efficiency, as it would allow for more optimisations than just solve this specific issue.

For example, both Agroal and Hibernate ORM perform #clearWarnings on the connection before returning it - I noticed it's actually not such a cheap operation on some drivers, even when there's no warnings to clear.

#### Have Agroal change design?

Agroal has chosen to implement this "Connection Wrapper" approach as a way to model the fact that each resource in an XA transaction needs to have reached the "end state" for the operations on each resource before the commit is allowed to proceed.

However, the usage of Hibernate ORM is not illegal per se: it is in a de-facto "end state" as it will no longer issue any statement on the connection - it simply hasn't really closed it yet.

This makes me think that perhaps the Connection Wrapper design in Agroal is too strict: there might be other ways to model the end state? For example, the wrapper could simply disallow any further use of the underlying connection rather than expecting it to have been closed already.

## But first

The ideas above all seem viable, and are not mutually exclusive. Personally I think I prefer @barreiro 's proposal to introduce a new before-commit release mode in ORM - and yet we could also optimise the Agroal/ORM integration by removing some more redundant checks.

And yet.. we probably need to consider first what are the implications of allowing people to use  `EntityManager.getTransaction()` : honestly that's confusing me a bit.

### PR DESCRIPTION
**Pull request by yrodiere**
Fixes #7242, #13273.

I ended up keeping support for `getTransaction()` for now, since it still works. We can open another pull request if we want to drop it.

Opening as draft because:

* The build is expected to fail for now; we need to upgrade to a version of ORM that includes https://github.com/hibernate/hibernate-orm/pull/3693 first.
* We need some additional testing as a previous version of this patch used to trigger deadlocks in the database. The fixes in ORM may solve these deadlocks, but that's absolutely not certain since we haven't found the exact cause yet. @johnaohara is trying to reproduce the deadlocks.

### CODE REVIEW

File: extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/session/JTASessionOptions.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.hibernate.resource.jdbc.spi.PhysicalConnectionHandlingMode;
+
+public class JTASessionOptions {
+    private static final String KEY = JTASessionOptions.class.getName();
+
+    public static void init(SessionFactory sessionFactory) {
+        SessionBuilder<?> options = sessionFactory.withOptions()
+                .autoClose(true) // .owner() is deprecated as well, so it looks like we need to rely on deprecated code...
+                .connectionHandlingMode(
+                        PhysicalConnectionHandlingMode.DELAYED_ACQUISITION_AND_RELEASE_BEFORE_TRANSACTION_COMPLETION)
```
Comment (Sanne) [inline extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/session/JTASessionOptions.java:15]: Could you explain to me again why these options have to be set again here?

Comment (yrodiere) [reply extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/session/JTASessionOptions.java:15]: Because managed, transaction-scoped sessions have different requirements from user-created sessions (direct calls to `createEntityManager()`) with manual transaction `begin`/`commit`.

Most important is the auto-close on transaction completion, which obviously only makes sense for transaction-scoped sessions. When a user calls `.createEntityManager()` manually, you'd expect them to close the EM manually.

For other settings, it's really about backward compatibility. From what I can see in `FastBootMetadataBuilder` we still allow custom settings for all sorts of properties (from `persistence.xml`?), and while other connection handling modes could make sense for manually-created entity managers, I don't think they would for transaction-scoped sessions.
This solution allows us to force these settings for transaction-scoped sessions, while preserving the ability to override them for manually-created entity managers.

To sum up, we could limit this to only setting `autoClose(true)` here, and assume that people setting the connection handling mode and other related settings manually know what they are doing.

**Comment by yrodiere** [comment]
@johnaohara gave the green light for merging this. I just rebased my branch. @Sanne , as soon as you're good with my answer to your review, we can go ahead and merge :)

Thanks to @johnaohara and @barreiro for looking into this!

**Comment by yrodiere** [comment]
There were a few problems with multi-tenancy, so I switched to a safer approach (and, I think, cleaner) approach in the last commit. It should be good now.

**Comment by yrodiere** [comment]
Fixed another test, apparently the metrics module was expecting two flushes for a single transaction. That's actually a problem that I fixed in this PR: we now correctly only flush once. I changed the test so that it passes.

Relevant commit: `Let Hibernate ORM handle the flush/close after a transaction`.

**Comment by Sanne** [comment]
> Fixed another test, apparently the metrics module was expecting two flushes for a single transaction. That's actually a problem that I fixed in this PR: we now correctly only flush once

Very nice!

File: extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootMetadataBuilder.java [hunk 2]
```diff
@@ -242,20 +242,18 @@ private MergedSettings mergeSettings(PersistenceUnitDescriptor persistenceUnit)
...
          */
-        {
-            final Object explicitSetting = cfg.get(AvailableSettings.CONNECTION_HANDLING);
-            if (explicitSetting == null) {
-                cfg.put(AvailableSettings.CONNECTION_HANDLING,
-                        PhysicalConnectionHandlingMode.DELAYED_ACQUISITION_AND_RELEASE_BEFORE_TRANSACTION_COMPLETION);
-            }
-        }
+        cfg.putIfAbsent(AvailableSettings.CONNECTION_HANDLING,
+                PhysicalConnectionHandlingMode.DELAYED_ACQUISITION_AND_RELEASE_BEFORE_TRANSACTION_COMPLETION);
```
Comment (Sanne) [inline extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootMetadataBuilder.java:256]: N.B. that semantics is not the same: the key can be defined while the value is set to null. No? Or is the change intentional?

Comment (yrodiere) [reply extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootMetadataBuilder.java:256]: If it's a problem here, it's probably also a problem just above, where we use `cfg.containsKey` to check whether a property is defined or not.

If we're going to worry about this, I think what we need to ask is: how would we end up in a situation where a property is set, but its value is `null`? From what I understand the only way for properties to be pre-set here is from a `persistence.xml`. I'm not sure if you can set something to `null` from `persistence.xml`?

**Comment by Sanne** [comment]
I'm not worried about it. I only wondered if you were aware of the subtle
change.

And I agree it probably doesn't matter in this case, as we only want to
support the option you're setting.

Perhaps we should enforce it, and use the containsKey method to trigger a
warning in case it was set.

FYI there's various undocumented ways in which this could get overridden,
e.g. hibernate.properties resources, environment variables, and
programmatically via the buildItem SPI. But while backdoors are possible
for exceptional cases they are not meant to be supported.

No need to change this, I just wanted to point out the subtlety for
awareness.

**Review by Sanne — APPROVED** [review]
Great work, thanks!
